pkg/storage/stores/shipper/compactor/index_set.go

```diff
@@ -0,0 +1,337 @@
+	"github.com/grafana/loki/pkg/storage/stores/shipper/storage"
+	shipper_util "github.com/grafana/loki/pkg/storage/stores/shipper/util"
+)
+
+const userIndexReadinessTimeout = 15 * time.Minute
+
+// indexSet helps with doing operations on a set of index files belonging to a single user or common index files shared by users.
+type indexSet struct {
+	ctx               context.Context
+	tableName, userID string
+	workingDir        string
+	baseIndexSet      storage.IndexSet
+	tableMarker       retention.TableMarker
+
+	compactedDBRecreated bool
+	uploadCompactedDB    bool
+	removeSourceObjects  bool
+
+	compactedDB   *bbolt.DB
+	sourceObjects []storage.IndexFile
+	logger        log.Logger
+	ready         chan struct{}
+	err           error
+}
+
+// newCommonIndex initializes a new index set for common index. It simply creates instance of indexSet without any processing.
+func newCommonIndex(ctx context.Context, tableName, workingDir string, compactedDB *bbolt.DB, uploadCompactedDB bool,
+	sourceFiles []storage.IndexFile, removeSourceFiles bool, baseCommonIndexSet storage.IndexSet, logger log.Logger) (*indexSet, error) {
+	if baseCommonIndexSet.IsUserIndexSet() {
+		return nil, fmt.Errorf("base index set is not for common index")
+	}
+
+	ui := indexSet{
+		ctx:                 ctx,
+		tableName:           tableName,
+		workingDir:          workingDir,
+		baseIndexSet:        baseCommonIndexSet,
+		compactedDB:         compactedDB,
+		uploadCompactedDB:   uploadCompactedDB,
+		sourceObjects:       sourceFiles,
+		removeSourceObjects: removeSourceFiles,
+		logger:              logger,
+		ready:               make(chan struct{}),
+	}
+	close(ui.ready)
+
+	return &ui, nil
+}
+
+// newUserIndex intializes a new index set for user index. Other than creating instance of indexSet, it also compacts down the source index.
+func newUserIndex(ctx context.Context, tableName, userID string, baseUserIndexSet storage.IndexSet, workingDir string, logger log.Logger) (*indexSet, error) {
+	if !baseUserIndexSet.IsUserIndexSet() {
+		return nil, fmt.Errorf("base index set is not for user index")
+	}
+
+	if err := util.EnsureDirectory(workingDir); err != nil {
+		return nil, err
+	}
+
+	ui := &indexSet{
+		ctx:          ctx,
+		tableName:    tableName,
+		userID:       userID,
+		workingDir:   workingDir,
+		baseIndexSet: baseUserIndexSet,
+		logger:       log.With(logger, "user-id", userID),
+		ready:        make(chan struct{}),
+	}
+
+	ui.initUserIndexSet(workingDir)
+	return ui, nil
+}
+
+// initUserIndexSet downloads the source index files and compacts them down to a single index file.
+func (is *indexSet) initUserIndexSet(workingDir string) {
+	defer close(is.ready)
+	ctx, cancelFunc := context.WithTimeout(is.ctx, userIndexReadinessTimeout)
+	defer cancelFunc()
+
+	is.sourceObjects, is.err = is.baseIndexSet.ListFiles(is.ctx, is.tableName, is.userID)
+	if is.err != nil {
+		return
+	}
+
+	compactedDBName := filepath.Join(workingDir, fmt.Sprint(time.Now().Unix()))
```

Comment:
Can this be conflincting across tenants if the init happen at the same second ?

Reply:
No, because the working directory for different users would be different. Each user would have its own working directory.